lib/teiserver/battle/balance/brute_force_avoid.ex

```diff
@@ -0,0 +1,161 @@
+  @moduledoc """
+  Overview:
+  Go through every possible combination and pick the best one. Each combination is given a score
+  Score = team rating penalty + broken avoid penalty + broken party penalty + broken max team rating diff penalty
+  team rating penalty = difference in team ratings
+  broken avoid penalty = broken avoid count * avoid importance
+  broken party penalty = broken party count * party importance
+  broken max team rating diff penalty = max team diff importance (if the team rating diff > @max_team_diff)
+
+  Only use for games with two teams.
+
+  This is not a balance algorithm that is callable players. It's a lib that can be used by another balance algorithm.
+  """
+  alias Teiserver.Battle.Balance.BruteForceAvoidTypes, as: BF
+  require Integer
+
+  @max_team_diff 17
+  @max_team_diff_importance 10000
+  @party_importance 1000
+  @avoid_importance 10
+
+  def get_best_combo(players, avoids, parties) do
+    potential_teams = potential_teams(length(players))
+    get_best_combo(potential_teams, players, avoids, parties)
+  end
+
+  @spec get_best_combo([integer()], [BF.player()], [[number()]], [[number()]]) ::
+          BF.combo_result()
+  def get_best_combo(combos, players, avoids, parties) do
+    players_with_index = Enum.with_index(players)
+
+    # Go through every possibility and get the combination with the lowest score
+    result =
+      Enum.map(combos, fn x ->
+        get_players_from_indexes(x, players_with_index)
+      end)
+      |> Enum.map(fn team ->
+        result = score_combo(team, players, avoids, parties)
+        Map.put(result, :first_team, team)
+      end)
+      |> Enum.min_by(fn z ->
+        z.score
+      end)
+
+    first_team = result.first_team
+
+    second_team =
+      players
+      |> Enum.filter(fn x ->
+        !Enum.any?(first_team, fn y ->
+          y.id == x.id
+        end)
+      end)
+
+    Map.put(result, :second_team, second_team)
+  end
+
+  @spec potential_teams(integer()) :: any()
+  def potential_teams(num_players) do
+    Teiserver.Helper.CombinationsHelper.get_combinations(num_players)
+  end
+
+  @spec score_combo([BF.player()], [BF.player()], [[number()]], [[number()]]) :: any()
+  def score_combo(first_team, all_players, avoids, parties) do
+    first_team_rating = get_team_rating(first_team)
+    both_team_rating = get_team_rating(all_players)
+    rating_diff_penalty = abs(both_team_rating - first_team_rating * 2)
+
+    max_team_diff_penalty =
+      cond do
+        rating_diff_penalty > @max_team_diff -> @max_team_diff_importance
+        true -> 0
+      end
+
+    # If max_team_diff_penalty is non zero don't even bother calculating avoid and party penalty
+    # since it's likely we'll discard this combo
+    {broken_avoid_penalty, broken_party_penalty} =
+      case max_team_diff_penalty do
+        0 ->
+          {count_broken_avoids(first_team, avoids) * @avoid_importance,
+           count_broken_parties(first_team, parties) * @party_importance}
```

Comment:
The more parties, the higher importance of this constraint. Did you consider dividing this by number of parties to get float 0-1 and only multiply that with importance factor? This way overall concept of parties has max bounded importance, same for avoids

Reply:
Hmm let me rethink this one.

Reply:
So there is a max bound due to these parameters
```
 @max_team_diff 14
  @max_team_diff_importance 10000
  ```
  If the rating diff is greater than @max_team_diff then parties and avoids will not be respected as those have loser importance